[Video][Subtitles] Use ffmpeg A53 sidedata instead of custom demuxer

enen92 opened the pull request:
## Description
While Kodi's closed caption decoder (although outdated when compared to upstream) is still far superior than that decoder contained within ffmpeg, the same can't be said about our custom demuxer implementation. There are a lot of reports/samples out there that simply cannot be played in Kodi but work just fine in mpv or ffplay. While investigating https://github.com/xbmc/xbmc/pull/22143 I realised the root cause of the problem was the fact our demuxed was extended from an initial support of `AV_CODEC_ID_MPEG2VIDEO` to `AV_CODEC_ID_H264` reusing some assumptions that only make sense in the former codec case (e.g. the start code). This code is complex and not maintained, ffmpeg seem to always fill the A53 side data with the closed caption data block, so we might want to use that instead in our demuxer (taking advantage of the already existing decoding code).

User feedback is appreciated...
@matthuisman would be great to know if this finally fixes the CC captions not available in some of your addons.

@ksooo it'd be nice if you could have a look into the code if you find some time. Looking for a pure code review (stuff to improve), I am not expecting you to be familiar with the inner-workings of the thing.

## Motivation and context
Improve subtitles and simplify our codebase at the same time. Get an alternative to the https://github.com/xbmc/xbmc/pull/22143 revert

## How has this been tested?
Runtime tested with all my CC samples, some that did work and others that didn't work. Those that used to not have captions detected now show the subtitle stream.

## What is the effect on users?
Hopefully better support for CC captions in Omega

## Screenshots (if appropriate):
**Before:**
<img width="774" alt="image" src="https://user-images.githubusercontent.com/7375276/209485351-57a468cd-6a68-447a-88c0-ff90a1d627b4.png">

**PR:**
<img width="778" alt="image" src="https://user-images.githubusercontent.com/7375276/209485287-4eeac0b0-f7dd-4681-bef1-1ce76386ff21.png">

<img width="798" alt="image" src="https://user-images.githubusercontent.com/7375276/209485268-b7973566-df91-4ca5-98e6-8e86576ae9ad.png">

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

enen92 committed:
[Video][Subtitles] Use ffmpeg A53 sidedata instead of custom demuxer

lrusak commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodec.h:
```diff
@@ -42,6 +42,31 @@ struct VideoPicture
+  *
+  * \param[in] sideData the data to store as A53 side data
+  */
+  void SetA53SideData(AVBufferRef* sideData) { m_A53SideData = sideData; }
```
Should we be calling `av_buffer_ref` and `av_buffer_unref` to do the ref counting for the buffer?

enen92 replied:
Yeah nice catch

garbear commented on xbmc/cores/VideoPlayer/Interface/CaptionBlock.h:
```diff
@@ -0,0 +1,67 @@
+    *
+    * \return the data stored in the closed caption block
+  */
+  std::vector<uint8_t> GetData() const { return m_data; }
```
This could be an expensive operation, as it always copies `m_data`. What about returning a const reference to m_data?

fuzzard commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -2886,6 +2857,34 @@ void CVideoPlayer::HandleMessages()
+            OpenDefaultStreams(false);
+          }
+          CDemuxStream* pSubStream = m_pCCDemuxer->GetStream(pkt->iStreamId);
+          if (pSubStream && m_CurrentSubtitle.id == pkt->iStreamId &&
+              m_CurrentSubtitle.source == STREAM_SOURCE_VIDEOMUX)
+            ProcessSubData(pSubStream, pkt);
+          else
+            CDVDDemuxUtils::FreeDemuxPacket(pkt);
+        }
```
Just wondering if theres going to be a leak from a "missing" FreeDemuxPacket. I tried to follow the ProcessSubData into the SendMessage, which (i think) goes into the VideoPlayer queue and i think is handled in https://github.com/xbmc/xbmc/blob/d3474a88f4a161101569f1c98bd866c17369fd6d/xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp#L302

Through all that, i can see in the process that it looks like it would hit https://github.com/xbmc/xbmc/blob/d3474a88f4a161101569f1c98bd866c17369fd6d/xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp#L528-L531, but theres no Free anywhere along the way.

Keep in mind, i really dont know the VP side of things, more of a query than anything.

fuzzard replied:
nm, the destructor of CDVDMsgDemuxerPacket has a free. all good :D

enen92 replied:
Yeah the code in this regard is exactly the same as before, I just nuked the Read() operation since that data is already provided by ffmpeg in the side data, so we just need to feed the decoder with that data.

Makes me question if the parse closed captions setting makes any sense after this

CastagnaIT commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodec.h:
```diff
@@ -42,6 +42,31 @@ struct VideoPicture
   void Reset(); // reinitialize members, videoBuffer will be released if set!
 
+  /*!
+  \brief Checks if the video picture has A53 side data (closed captions).
```
i know you hate descriptions 😂
but if better if you can fix them all

enen92 replied:
Hum, what's wrong with this doxy block? The missing * before the brief keyword?

CastagnaIT replied:
yes right like this
![immagine](https://user-images.githubusercontent.com/3257156/209700884-82ff4467-3290-4317-96c8-7ba87347066d.png)

CastagnaIT commented on xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp:
```diff
@@ -733,6 +734,18 @@ bool CVideoPlayerVideo::ProcessDecoderOutput(double &frametime, double &pts)
       pts += m_picture.iDuration * m_speed / abs(m_speed);
 
+    // ATSC A53 Closed Captions
+    if (CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+            CSettings::SETTING_SUBTITLES_PARSECAPTIONS) &&
+        m_picture.HasA53SideData())
```
can you reverse the content of "if" statement?
so that settings interface is called only when a53 data exists
```suggestion
    if (m_picture.HasA53SideData() && ServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_SUBTITLES_PARSECAPTIONS))
```

enen92 commented:
@CastagnaIT thanks for the tips, hopefully done right on the last push

lrusak commented on xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodec.h:
```diff
@@ -42,6 +42,31 @@ struct VideoPicture
+  * \brief Gets the A53 side data of the video picture (closed captions)
+  * \return the A53 side data, reseting the value
+  */
+  AVBufferRef* ConsumeA53SideData()
```
Shoud this be `GetA53SideData` ? As you say that in the doxy.

lrusak replied:
It might also make sense to have a `ReleaseA53SideData` method that calls the `av_buffer_unref` that way the ref and unref are contained.

enen92 replied:
I opted for the consume Vs get because it's set as nullptr. It'd be better to just use smart pointers here and move the pointer but I couldn't manage to do it

enen92 replied:
@lrusak if you are bored enough and could help with this it'd be appreciated. I'd say the proper way to do this is to store a unique_ptr for the side data in this class instead of a raw pointer and move it when calling get, so that it's set to nullptr internally in the class. Moreover I think it could in theroty have a custom deleter that would unref the buffer on destruction, avoiding the callee to call into unref manually.
Unfortunately I couldn't manage to do it because of the default copy constructors we have on VideoPicture if I recall correctly.

lrusak commented on xbmc/cores/VideoPlayer/DVDMessage.h:
```diff
@@ -331,3 +333,34 @@ class CDVDMsgSubtitleClutChange : public CDVDMsg
+  /*!
+  * \brief Default destructor
+  */
+  ~CDVDMsgSubtitleCCData() = default;
```
```suggestion
  ~CDVDMsgSubtitleCCData() override = default;
```

lrusak commented on xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp:
```diff
@@ -733,6 +734,18 @@ bool CVideoPlayerVideo::ProcessDecoderOutput(double &frametime, double &pts)
+        CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(
+            CSettings::SETTING_SUBTITLES_PARSECAPTIONS))
+    {
+      auto a53SideData = m_picture.ConsumeA53SideData();
```
I would say that this usage of auto isn't that great as the type isn't know from the method.
```suggestion
      AVBufferRef* a53SideData = m_picture.ConsumeA53SideData();
```

lrusak commented on xbmc/cores/VideoPlayer/VideoPlayerVideo.cpp:
```diff
@@ -733,6 +734,18 @@ bool CVideoPlayerVideo::ProcessDecoderOutput(double &frametime, double &pts)
+            CSettings::SETTING_SUBTITLES_PARSECAPTIONS))
+    {
+      auto a53SideData = m_picture.ConsumeA53SideData();
+      std::unique_ptr<CCaptionBlock> ccData =
```
This is however a good place to use auto as you know it will be `std::unique_ptr<CCaptionBlock>` from the assignment.
```suggestion
      auto ccData =
```

lrusak commented on xbmc/cores/VideoPlayer/VideoPlayer.cpp:
```diff
@@ -2886,6 +2857,34 @@ void CVideoPlayer::HandleMessages()
+              m_SelectionStreams.CountTypeOfSource(STREAM_SUBTITLE, STREAM_SOURCE_VIDEOMUX))
+          {
+            m_SelectionStreams.Clear(STREAM_SUBTITLE, STREAM_SOURCE_VIDEOMUX);
+            m_SelectionStreams.Update(NULL, m_pCCDemuxer.get(), "");
```
```suggestion
            m_SelectionStreams.Update(nullptr, m_pCCDemuxer.get(), "");
```

lrusak commented on xbmc/cores/VideoPlayer/Interface/CaptionBlock.h:
```diff
@@ -0,0 +1,67 @@
+   *
+   * \return the size of the closed caption block
+  */
+  unsigned long Size() const { return m_data.size(); }
```
I don't see this actually used anywhere?

enen92 committed:
[video][subtitles] Use unique_ptr for A53 side data

enen92 commented:
@lrusak think I've managed to use a unique_ptr for the side data (added on a separate commit). Please take a look if you have a chance.

enen92 committed:
[Closed Captions] Discard duplicate dual codes

enen92 commented:
Added a new commit to fix another issue, according to the standard duplicate dual codes must be discarded. Also in ccextractor:
https://github.com/CCExtractor/ccextractor/blob/master/src/lib_ccx/ccx_decoders_608.c#L1193-L1202

The checking code was already there but was only being applied to the control chars, not regular dual char codes.

**Before:**
![image](https://user-images.githubusercontent.com/7375276/210350861-22cb6f2d-0ea4-4beb-9179-b573fc60110e.png)


**After:**
![image](https://user-images.githubusercontent.com/7375276/210350764-6e3b570f-0292-4263-8edc-423fb9fb843b.png)

emveepee commented:
@enen92 is this for subtitles as shown in the title or closed captioning 608/708 (or both)?

enen92 commented:
Closed captions (eia-608/708) which are also subtitles.
This is still a bit WIP, I've found one sample that didn't work and I'm trying to figure out the issue.

emveepee commented:
> Closed captions (eia-608/708) which are also subtitles. This is still a bit WIP, I've found one sample that didn't work and I'm trying to figure out the issue.
> 
> If you have content and are able to give them a go it'd be great.

Here in North America we generally use the term closed captioning for accessibility (it also has description of sounds, music playing  notes etc) and subtitles for alternate languages which is why I wasn't sure.

I will add this after Nexus final and try with Live TV we have 608/708.   I have a client for Android too and Exoplayer does a better job then Kodi so I will use that for comparing.  The Exoplayer code is here
https://github.com/google/ExoPlayer/tree/release-v2/library/extractor/src/main/java/com/google/android/exoplayer2/text/cea   It doesn't work with as many samples as ccextractor but it might help you too

enen92 committed:
[subtitles][closed captions] Remove unneeded ccReorderbuffer

enen92 committed:
[VideoPlayer][Subtitles] Remove codec restrictions for closed captions

ffmpeg already sets the sidedata whenever it is detected and as a
result there is no need to restrict this to MPEG or H264. Furthermore
other codecs also support EIA608 (e.g. HEVC)

enen92 commented:
I am interested in samples or accessing addon streams (temporarily) which the closed captions cannot be displayed/demuxed in Kodi (I mean no subtitles at all not exactly errors displaying, colouring or positioning the subtitle entries).

So far the only public samples that didn't work (also didn't work before) are those provided by the dash industry forum (e.g: https://livesim.dashif.org/dash/vod/testpic_2s/cea608.mpd) but unfortunately they are completely broken and can't be used as a reference for now, see:

https://github.com/Dash-Industry-Forum/dash-live-source-simulator/issues/99#issuecomment-1374546849
https://github.com/Dash-Industry-Forum/dash-live-source-simulator/issues/115

Maybe we can give them a go later on if those major issues are fixed

enen92 commented:
Added a few more commits and I consider this finished:

- Simplify the code of ccdemuxer as we don't need two buffers anymore
- Removed the codec restrictions we used to have for calling the cc demuxer. FFMpeg sets the side data whenever it is detected and other codecs might indeed have closed captions (e.g. HEVC not only mpeg && h264 as before) and the code path is the same.
- Decided to remove our old setting "Enable parse of closed captions" since the extraction of closed captions no longer puts any additional load on the CPU. It's done by ffmpeg when decoding the pictures anyway. Any load caused by decoding the captions is pretty similar to any other subtitle type.

Also the DASH foundation has fixed their public streams.

I don't want to add more stuff into this (nor any other fixes out of the scope of CC demuxing) to avoid exploding the diff. Plan to get this merged by the end of the month unless someone objects. Getting a green and some testing would still be great.

Note CC decoding and general behaviour is (and should be) still pretty much the same. Any issues with colors, positions, some entries no showing are out of the scope of this PR. I've just changed the way we get the data.

emveepee commented:
I've been trying this with live TV today and in general it seems like a definite improvement.  Omega is getting less stable for general use though so I think my testing will be limited to your PR I can't use master at the moment..   

You probably should look at this sample since the captioning just stops https://www.mediafire.com/file/gefs76f1ooppj3o/subtitles.ts/file but can be re-enabled.  Musical notes show as boxes too, don't know if that is in scope.  

It is also hard to compare it to other players with positional support especially when several people are talking but it still seems like an improvement.   I don't need cc's but I am interested in it for the community.  It is really nice not to have to configure Kodi to parse them.

CastagnaIT commented:
i have tested with my samples and all worked good job!

However, I think a solution should be found for the removed CC setting
was here because also for another reason, so due to track language problem,
currently we have no way to know CC language type and this lead to following situation:

for example:
you play a video that have CC Spanish,
the user want display only Italian subtitles (then set kodi sub language setting to Italian)
but since the CC track have unknown language, kodi player always display CC subtitles no matter what language you set in kodi language setting, so without a setting the user always have to manually disable subtitles each time that play a video

enen92 commented:
Thanks for the feedback. Yeap I am not a big fan of it either, they are always enabled by default and their definition is too generic. I have to take a look to into the standards to check if there's any way to flag some of the properties (language, type, etc) but I don't think so. Also not sure if inputsream add-ons can create those streams beforehand since they are usually flagged in the respective manifests...but then we also lack anyway to create multiple streams of the same type (e.g. 2 cea608 streams)...
I guess I'll just remove the setting removal commit for now to keep the previous behaviour (this is about the demuxer anyway). There's lot of room for future improvements.

enen92 committed:
[Subtitles] Remove AVCodec from ccDemuxer - no longer needed

CastagnaIT reviewed (approved):
generically after the test it for me its fine,
but i cannot evaluate the changes on VP

emveepee commented:
@enen92 not having to enable parsing was one of the major benefits of this PR.  I agree that they should not be forced on but couldn't there be a compromise be made to not display them by default like other subtitle types?  In any case for closed captioning the language cannot be consider safe, here often the language identifier when it is used and not unknown is wrong.

enen92 commented:
Yes, later...
This is not really the focus of this PR. I agree with castagna we should have a proper mechanism in place if we decide to drop the setting.
